File: _290_WordPattern.java
// space complexity - o(2n)
    	//time - o(n)
      //Ran on leeetcode successfully : Yes
      // Problem faced  : No
    	//Approach : match each charater with other, from s to t and from t to s


class Solution {
    public boolean wordPattern(String pattern, String s) {

        Map<Character, String> map1 = new HashMap<>();
        Map<String, Character> map2 = new HashMap<>();

        int start = 0;
        int end = 0;
        for(int i = 0; i< pattern.length(); i++){
            Character ch = pattern.charAt(i);

            while(end < s.length() && s.charAt(end) != ' '){
                end++;
            }
            String word = s.substring(start,end);

            if(map1.containsKey(ch) && !map1.get(ch).equals(word)) return false;
            map1.put(ch, word);

            if(map2.containsKey(word) && !(map2.get(word) == ch)) return false;
            map2.put(word, ch);

            if(end == s.length() && i != pattern.length() -1) return false;

            end++;
            start = end;
        }
        if(end < s.length()) return false;
        return true;
    }
}
//_290_WordPattern

File: 290-WordPattern.py
#Time Complexity : O(N ) where N is number of words in s or number of characters in pattern
# Space Complexity : O(N) where N represents unique words in s 
# Did this code successfully run on Leetcode : yes
# Any problem you faced while coding this :
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        hashmap1 = {}
        hashmap2 = {}
        strList = s.split()
        if len(pattern)  != len(strList):
            return False
        for i in range(0,len(pattern)):
            if pattern[i] in hashmap1:
                if hashmap1[pattern[i]] != strList[i]:
                    return False
            else:
                hashmap1[pattern[i]] = strList[i]
            
            if strList[i] in hashmap2:
                if hashmap2[strList[i]] != pattern[i]:
                    return False
            else:
                hashmap2[strList[i]] = pattern[i]      
        return True

